substrate/frame/delegated-staking/src/types.rs

```diff
@@ -0,0 +1,308 @@
+		let ledger = AgentLedger::<T>::get(agent).ok_or(Error::<T>::NotAgent)?;
+		Ok(Agent { key: agent.clone(), ledger })
+	}
+
+	/// Remove funds that are withdrawn from [Config::CoreStaking] but not claimed by a delegator.
+	///
+	/// Checked decrease of delegation amount from `total_delegated` and `unclaimed_withdrawals`
+	/// registers. Consumes self and returns a new instance of self if success.
+	pub(crate) fn remove_unclaimed_withdraw(
+		self,
+		amount: BalanceOf<T>,
+	) -> Result<Self, DispatchError> {
+		let new_total_delegated = self
+			.ledger
+			.total_delegated
+			.checked_sub(&amount)
+			.defensive_ok_or(ArithmeticError::Overflow)?;
+		let new_unclaimed_withdrawals = self
+			.ledger
+			.unclaimed_withdrawals
+			.checked_sub(&amount)
+			.defensive_ok_or(ArithmeticError::Overflow)?;
+
+		Ok(Agent {
+			ledger: AgentLedger {
+				total_delegated: new_total_delegated,
+				unclaimed_withdrawals: new_unclaimed_withdrawals,
+				..self.ledger
+			},
+			..self
+		})
+	}
+
+	/// Add funds that are withdrawn from [Config::CoreStaking] to be claimed by delegators later.
+	pub(crate) fn add_unclaimed_withdraw(
+		self,
+		amount: BalanceOf<T>,
+	) -> Result<Self, DispatchError> {
+		let new_unclaimed_withdrawals = self
+			.ledger
+			.unclaimed_withdrawals
+			.checked_add(&amount)
+			.defensive_ok_or(ArithmeticError::Overflow)?;
+
+		Ok(Agent {
+			ledger: AgentLedger { unclaimed_withdrawals: new_unclaimed_withdrawals, ..self.ledger },
+			..self
+		})
+	}
+
+	/// Amount that is delegated but not bonded yet.
+	///
+	/// This importantly does not include `unclaimed_withdrawals` as those should not be bonded
+	/// again unless explicitly requested.
+	pub(crate) fn available_to_bond(&self) -> BalanceOf<T> {
+		let bonded_stake = self.bonded_stake();
+		let stakeable = self.ledger.stakeable_balance();
+
+		defensive_assert!(
+			stakeable >= bonded_stake,
+			"cannot be bonded with more than total amount delegated to agent"
+		);
+
+		stakeable.saturating_sub(bonded_stake)
+	}
+
+	/// Remove slashes from the `AgentLedger`.
+	pub(crate) fn remove_slash(self, amount: BalanceOf<T>) -> Self {
+		let pending_slash = self.ledger.pending_slash.defensive_saturating_sub(amount);
+		let total_delegated = self.ledger.total_delegated.defensive_saturating_sub(amount);
+
+		Agent { ledger: AgentLedger { pending_slash, total_delegated, ..self.ledger }, ..self }
+	}
+
+	/// Get the total stake of agent bonded in [`Config::CoreStaking`].
+	pub(crate) fn bonded_stake(&self) -> BalanceOf<T> {
+		T::CoreStaking::total_stake(&self.key).unwrap_or(Zero::zero())
+	}
+
+	/// Returns true if the agent is bonded in [`Config::CoreStaking`].
+	pub(crate) fn is_bonded(&self) -> bool {
+		T::CoreStaking::stake(&self.key).is_ok()
+	}
+
+	/// Returns the reward account registered by the agent.
+	pub(crate) fn reward_account(&self) -> &T::AccountId {
+		&self.ledger.payee
+	}
+
+	/// Save self to storage.
+	pub(crate) fn save(self) {
```

Comment:
How about having a `checked_mutate` like fn here that does all the integrity checks before saving to storage? some checks could be:

- `stakeable >= bonded_stake`
- `pending_slash <= bonded_stake`
- etc..

Reply:
Ah, I see that `fn save_or_kill` below implements a similar approach before killing the ledger. I'd check the more broad ledger integrity checks here and fail with `BadState` if something is off. 

If we get into a similar situation with the corrupted staking ledger has we saw recently, at least we don't corrupt the ledger further.